### PR TITLE
feat: supported Whitespace Accumulator Interceptor

### PR DESCRIPTION
**Pull request by tg-epam**
Adds an interceptor that batches whitespace-only stream chunks and flushes them on content or timeout.

### CODE REVIEW

File: aidial_interceptors_sdk/examples/chat_completion/whitespace_accumulator.py [hunk 1]
```diff
@@ -0,0 +1,87 @@
...
+
+
+class WhitespaceAccumulatorInterceptor(ChatCompletionInterceptor):
+    buffer: list[str] = []
+    start: float = 0.0
+    timeout: float = _FLUSH_TIMEOUT
+    send_content = False
+
+    def flush(self) -> str:
+        _log.debug(f"flushed {len(self.buffer)} elements")
```
Comment (adubovik) [inline aidial_interceptors_sdk/examples/chat_completion/whitespace_accumulator.py]: 1. Avoid debug prints that generate too much content and are therefore not very helpful.
This one will literally print one line for every single streaming chunk on a happy path.
Let's replace it with info log when the interceptor does something non-trivial:

```py
if len(self.buffer) > 1:
    _log.info(...)
```

2. `elements` (implementation details) -> `chunks` (notion from the chat completion API)

File: aidial_interceptors_sdk/examples/chat_completion/whitespace_accumulator.py [hunk 2]
```diff
@@ -0,0 +1,87 @@
...
+            self.set_content(chunk, flushed + content)
+            self.send_chunk(chunk)
+            self.send_content = True
+        else:
+            _log.debug("buffered content")
+            self.buffer.append(content)
+
+    @override
+    async def on_stream_end(self) -> None:
+        if len(self.buffer) == 0:
```
Comment (adubovik) [inline aidial_interceptors_sdk/examples/chat_completion/whitespace_accumulator.py]: `if not self.buffer:`

File: aidial_interceptors_sdk/examples/chat_completion/whitespace_accumulator.py [hunk 3]
```diff
@@ -0,0 +1,87 @@
...
+            return
+        flushed = self.flush()
+        chunk = ContentChunk(flushed, 0)
+        if self.send_content is True:
```
Comment (adubovik) [inline aidial_interceptors_sdk/examples/chat_completion/whitespace_accumulator.py]: `if self.send_content:`

File: README.md [hunk 4]
```diff
@@ -128,6 +128,7 @@ They are provided solely as examples to demonstrate basic use cases of intercept
 |replicator:N|Generic|Calls the upstream N times and combines the N response into a single response. Could be useful for stabilization of model's output, since certain models aren't deterministic.|
 |cache|Generic|Caches incoming chat completion requests. **Not ready for production use. Use at your discretion**|
 |no-op|Generic|No-op interceptor - does not modify the request or the response, simply proxies the upstream|
+|whitespace-accumulator|Generic|Buffers whitespace-only stream chunks and flushes them together when non-whitespace content arrives or a timeout is reached to reduce streaming noise.|
```
Comment (adubovik) [inline README.md]: List known limitations - the interceptor works only for the first response choice.

File: aidial_interceptors_sdk/examples/chat_completion/whitespace_accumulator.py [hunk 5]
```diff
@@ -0,0 +1,87 @@
...
+
+_log = logging.getLogger(__name__)
+
+_FLUSH_TIMEOUT = 1.0
+
+
+class WhitespaceAccumulatorInterceptor(ChatCompletionInterceptor):
+    buffer: list[str] = []
+    start: float = 0.0
+    timeout: float = _FLUSH_TIMEOUT
```
Comment (adubovik) [inline aidial_interceptors_sdk/examples/chat_completion/whitespace_accumulator.py]: Let's add the unit - `timeout_sec`

File: aidial_interceptors_sdk/examples/chat_completion/whitespace_accumulator.py [hunk 6]
```diff
@@ -0,0 +1,87 @@
...
+        return content
+
+    def need_flush(self, data: str) -> bool:
+        passed = time.perf_counter() - self.start
+        return not data.isspace() or (passed > self.timeout)
+
+    def get_content(self, chunk: dict) -> str:
+        try:
+            return (
+                chunk["choices"][0]["delta"].get("content")
```
Comment (adubovik) [inline aidial_interceptors_sdk/examples/chat_completion/whitespace_accumulator.py]: 1. If `delta` is missing from `chunk["choices"][0]`, then the expression blows up without giving the chance for the alternative to execute: `chunk["choices"][0]["message"].get("content")` 
2. (1) doesn't actually matter, because the interceptor SDK always provides "delta" in the response chunks. That is, you don't need to deal with "message", only with "delta".

File: tests/utils/applications.py [hunk 7]
```diff
@@ -16,6 +16,18 @@ async def chat_completion(
             choice.append_content(request.messages[-1].text())
 
 
+class WhitespaceEchoApplication(ChatCompletion):
+    async def chat_completion(
+        self, request: Request, response: Response
+    ) -> None:
+        with response.create_single_choice() as choice:
+            for _ in range(10):
+                choice.append_content(" ")
```
Comment (adubovik) [inline tests/utils/applications.py]: Send also 10 empty strings before and after non-empty content, and see what happens. They should be merged too.

File: aidial_interceptors_sdk/examples/chat_completion/whitespace_accumulator.py [hunk 8]
```diff
@@ -0,0 +1,87 @@
...
+)
+
+_log = logging.getLogger(__name__)
+
+_FLUSH_TIMEOUT = 1.0
+
+
+class WhitespaceAccumulatorInterceptor(ChatCompletionInterceptor):
+    buffer: list[str] = []
+    start: float = 0.0
```
Comment (adubovik) [inline aidial_interceptors_sdk/examples/chat_completion/whitespace_accumulator.py]: It's a misnomer. It's rather `last_send`.

File: aidial_interceptors_sdk/examples/chat_completion/whitespace_accumulator.py [hunk 9]
```diff
@@ -0,0 +1,87 @@
...
+    timeout: float = _FLUSH_TIMEOUT
+    send_content = False
```
Comment (adubovik) [inline aidial_interceptors_sdk/examples/chat_completion/whitespace_accumulator.py]: Let's rename to `any_chunk_sent`.
`send_content` sounds like a command rather than a predicate.

File: aidial_interceptors_sdk/examples/chat_completion/whitespace_accumulator.py [hunk 10]
```diff
@@ -0,0 +1,87 @@
...
+
+    def flush(self) -> str:
+        _log.debug(f"flushed {len(self.buffer)} elements")
+        content = "".join(self.buffer)
+        self.buffer.clear()
+        self.start = time.perf_counter()
+        return content
+
+    def need_flush(self, data: str) -> bool:
```
Comment (adubovik) [inline aidial_interceptors_sdk/examples/chat_completion/whitespace_accumulator.py]: Why `data` here and everywhere else `content`? Let's rename to `new_content`.

File: tests/utils/applications.py [hunk 11]
```diff
@@ -16,6 +16,18 @@ async def chat_completion(
             choice.append_content(request.messages[-1].text())
 
 
+class WhitespaceEchoApplication(ChatCompletion):
+    async def chat_completion(
+        self, request: Request, response: Response
+    ) -> None:
+        with response.create_single_choice() as choice:
```
Comment (adubovik) [inline tests/utils/applications.py:28]: 1. Add the test with just a single empty string sent (`""`)
2. Add the test with just a single whitespace string sent (e.g. `" "`)

File: aidial_interceptors_sdk/examples/chat_completion/whitespace_accumulator.py [hunk 12]
```diff
@@ -0,0 +1,87 @@
...
+
+    @override
+    async def on_stream_start(self) -> None:
+        self.start = time.perf_counter()
+
+    @override
+    async def on_stream_chunk(self, chunk: dict) -> None:
+        content = self.get_content(chunk)
+        if content == "":
+            self.send_chunk(chunk)
```
Comment (adubovik) [inline aidial_interceptors_sdk/examples/chat_completion/whitespace_accumulator.py]: `self.send_content = True` here too?

File: aidial_interceptors_sdk/examples/chat_completion/whitespace_accumulator.py [hunk 13]
```diff
@@ -0,0 +1,87 @@
...
+            return
+
+        if self.need_flush(content):
+            flushed = self.flush()
+            self.set_content(chunk, flushed + content)
+            self.send_chunk(chunk)
+            self.send_content = True
+        else:
+            _log.debug("buffered content")
```
Comment (adubovik) [inline aidial_interceptors_sdk/examples/chat_completion/whitespace_accumulator.py]: This is going to trigger a lot of logs, especially for code generation. Let's omit it.

File: aidial_interceptors_sdk/examples/chat_completion/whitespace_accumulator.py [hunk 14]
```diff
@@ -0,0 +1,87 @@
...
+
+    def get_content(self, chunk: dict) -> str:
+        try:
+            return (
+                chunk["choices"][0]["delta"].get("content")
+                or chunk["choices"][0]["message"].get("content")
+                or ""
+            )
+        except (KeyError, IndexError, TypeError):
+            return ""
```
Comment (adubovik) [inline aidial_interceptors_sdk/examples/chat_completion/whitespace_accumulator.py]: You are conflating three cases here:
1. `chunk["choices"][0]["message"].get("content")` is equal to ""
2. `chunk["choices"][0]["message"].get("content")` is equal to None
3. `chunk["choices"][0]["message"].get("content")` triggers an exception

Should they all be handled the same way?

File: aidial_interceptors_sdk/examples/chat_completion/whitespace_accumulator.py [hunk 15]
```diff
@@ -0,0 +1,87 @@
...
+            return
+
+        if self.need_flush(content):
+            flushed = self.flush()
+            self.set_content(chunk, flushed + content)
+            self.send_chunk(chunk)
+            self.send_content = True
+        else:
+            _log.debug("buffered content")
+            self.buffer.append(content)
```
Comment (adubovik) [inline aidial_interceptors_sdk/examples/chat_completion/whitespace_accumulator.py:89]: We are actually loosing non-content related information in the `chunk`.

**Review by adubovik — CHANGES_REQUESTED** [review]

File: tests/unit_tests/examples/test_whitespace.py [hunk 16]
```diff
@@ -0,0 +1,86 @@
...
+from tests.utils.dial_app import create_openai_client
+from tests.utils.json import has_type, match_objects, memorize
+
+
+@pytest.mark.parametrize("stream", [False, True])
+def test_whitespace_accumulator_interceptor(stream: bool):
+    spaces = " " * 10
+
+    def callback(request: Request, response: Response, choice: Choice):
+        # choice.append_content("")
```
Comment (adubovik) [inline tests/unit_tests/examples/test_whitespace.py]: Avoid commented-out code

File: tests/unit_tests/examples/test_whitespace.py [hunk 17]
```diff
@@ -0,0 +1,86 @@
...
+    )
+
+    if isinstance(response, ChatCompletion):
+        match_objects(
+            response.to_dict(),
+            checker(
+                {
+                    "role": "assistant",
+                    "content": f"{spaces}hello{spaces}",
+                    # "content": f"",
```
Comment (adubovik) [inline tests/unit_tests/examples/test_whitespace.py]: Avoid commented-out code

File: tests/unit_tests/examples/test_whitespace.py [hunk 18]
```diff
@@ -0,0 +1,86 @@
...
+                },
+                finish_reason="stop",
+            ),
+        )
+
+    else:
+        expected_chunks = [
+            checker({"content": f"{spaces}hello", "role": "assistant"}),
+            checker({"content": f"{spaces}"}, finish_reason="stop"),
+            # checker({"content": "", "role": "assistant"}, finish_reason="stop"),
```
Comment (adubovik) [inline tests/unit_tests/examples/test_whitespace.py]: Avoid commented-out code

File: tests/unit_tests/examples/test_whitespace.py [hunk 19]
```diff
@@ -0,0 +1,86 @@
...
+        ]
+
+        for actual, expected in itertools.zip_longest(
+            response, expected_chunks
+        ):
+            print(f"expected {expected}")
```
Comment (adubovik) [inline tests/unit_tests/examples/test_whitespace.py]: Avoid debug prints in tests

File: aidial_interceptors_sdk/examples/chat_completion/whitespace_accumulator.py [hunk 20]
```diff
@@ -0,0 +1,64 @@
...
+        except (KeyError, IndexError, TypeError):
+            return None
+
+    @override
+    async def on_stream_start(self) -> None:
+        self.last_send = time.perf_counter()
+
+    @override
+    async def on_stream_chunk(self, chunk: dict) -> None:
+        if not self.cur_chunk:
```
Comment (adubovik) [inline aidial_interceptors_sdk/examples/chat_completion/whitespace_accumulator.py]: Why is the first chunk **always** buffered? It shouldn't be if it isn't a whitespace chunk.

File: aidial_interceptors_sdk/examples/chat_completion/whitespace_accumulator.py [hunk 21]
```diff
@@ -0,0 +1,64 @@
...
+            self.cur_chunk = copy.deepcopy(chunk)
+            self.cur_chunks += 1
+            return
+
+        merge(self.cur_chunk, chunk)
```
Comment (adubovik) [inline aidial_interceptors_sdk/examples/chat_completion/whitespace_accumulator.py]: Here you are running into the same performance issue we are trying to solve with this interceptor.

It's worth adding a dedicated performance test.

File: tests/unit_tests/examples/test_whitespace.py [hunk 22]
```diff
@@ -0,0 +1,86 @@
...
+    WhitespaceAccumulatorInterceptor,
+)
+from tests.utils.applications import WhitespaceEchoApplication
+from tests.utils.chunks import create_chunk_checker
+from tests.utils.dial_app import create_openai_client
+from tests.utils.json import has_type, match_objects, memorize
+
+
+@pytest.mark.parametrize("stream", [False, True])
+def test_whitespace_accumulator_interceptor(stream: bool):
```
Comment (adubovik) [inline tests/unit_tests/examples/test_whitespace.py]: Add a test with chunks with empty `content` fields and non-empty other fields (like `custom_content`).

File: aidial_interceptors_sdk/examples/chat_completion/whitespace_accumulator.py [hunk 23]
```diff
@@ -0,0 +1,64 @@
...
+    def flush(self) -> None:
+        assert self.cur_chunk is not None
+        _log.info(f"flushed {self.cur_chunks} chunks")
+        self.last_send = time.perf_counter()
+        self.send_chunk(self.cur_chunk)
+        self.cur_chunk = None
+        self.cur_chunks = 0
+
+    def need_flush(self) -> bool:
+        assert self.cur_chunk is not None
```
Comment (adubovik) [inline aidial_interceptors_sdk/examples/chat_completion/whitespace_accumulator.py]: Pass `cur_chunk` as an argument to avoid the assert.

File: aidial_interceptors_sdk/examples/chat_completion/whitespace_accumulator.py [hunk 24]
```diff
@@ -0,0 +1,64 @@
...
+
+
+class WhitespaceAccumulatorInterceptor(ChatCompletionInterceptor):
+    cur_chunk: dict | None = None
+    last_send: float = 0.0
+    cur_chunks: int = 0
+    timeout_sec: float = _FLUSH_TIMEOUT
+
+    def flush(self) -> None:
+        assert self.cur_chunk is not None
```
Comment (adubovik) [inline aidial_interceptors_sdk/examples/chat_completion/whitespace_accumulator.py]: Pass `cur_chunk` as an argument to avoid the assert.

**Review by adubovik — CHANGES_REQUESTED** [review]

File: aidial_interceptors_sdk/examples/chat_completion/whitespace_accumulator.py (hunk 24 above)
Comment (tg-epam) [reply aidial_interceptors_sdk/examples/chat_completion/whitespace_accumulator.py]: This is invariant of the class, so I don't see a need to do this

File: aidial_interceptors_sdk/examples/chat_completion/whitespace_accumulator.py [hunk 25]
```diff
@@ -0,0 +1,114 @@
...
+            chunk["choices"][0]["delta"]["content"] = content
+        except (KeyError, IndexError, TypeError):
+            return None
+
+    def set_chunk_if_needed(self, chunk: dict):
+        if not self.cur_chunk:
+            self.cur_chunk = chunk
+
+    def match_cache_structure(self, chunk: dict) -> bool:
+        if not ("choices" in chunk.keys()):
```
Comment (adubovik) [inline aidial_interceptors_sdk/examples/chat_completion/whitespace_accumulator.py]: simplify:

```py
if (choices := chunk.get("choices")) is None:
    return False
```

File: aidial_interceptors_sdk/examples/chat_completion/whitespace_accumulator.py [hunk 26]
```diff
@@ -0,0 +1,114 @@
...
+
+
+class WhitespaceAccumulatorInterceptor(ChatCompletionInterceptor):
+    last_send: float = 0.0
+    cur_chunk: dict | None = None
+    cur_chunks: int = 0
+    timeout_sec: float = _FLUSH_TIMEOUT
+    buffer: list[str] = []
+
+    def save_to_cache(self, chunk: dict):
```
Comment (adubovik) [inline aidial_interceptors_sdk/examples/chat_completion/whitespace_accumulator.py]: the method is unused

File: aidial_interceptors_sdk/examples/chat_completion/whitespace_accumulator.py [hunk 27]
```diff
@@ -0,0 +1,114 @@
...
+        try:
+            chunk["choices"][0]["delta"]["content"] = content
+        except (KeyError, IndexError, TypeError):
+            return None
+
+    def set_chunk_if_needed(self, chunk: dict):
+        if not self.cur_chunk:
+            self.cur_chunk = chunk
+
+    def match_cache_structure(self, chunk: dict) -> bool:
```
Comment (adubovik) [inline aidial_interceptors_sdk/examples/chat_completion/whitespace_accumulator.py]: The interceptor has nothing to do with **caching**.

I suggest merging `match_cache_structure` and `get_content` into a single parsing method 

```py
def parse_content_only_chunk(chunk: dict) -> str | None:
    ...
    return content
```

File: aidial_interceptors_sdk/examples/chat_completion/whitespace_accumulator.py [hunk 28]
```diff
@@ -0,0 +1,114 @@
...
+        delta = choice["delta"]
+        if not isinstance(delta, dict) or not set(delta.keys()).issubset(
+            {
+                "content",
+                "role",
+            }
+        ):
+            return False
+
+        finish_reason = choice["finish_reason"]
```
Comment (adubovik) [inline aidial_interceptors_sdk/examples/chat_completion/whitespace_accumulator.py]:
```py
        if choice["finish_reason"]:
            return False
```

Comment (adubovik) [inline aidial_interceptors_sdk/examples/chat_completion/whitespace_accumulator.py]: Move this shortcut right after this condition:

```py
        if not isinstance(choice, dict) or set(choice.keys()) != {
            "delta",
            "index",
            "finish_reason",
        }:
            return False
```

File: aidial_interceptors_sdk/examples/chat_completion/whitespace_accumulator.py [hunk 29]
```diff
@@ -0,0 +1,114 @@
...
+        if not isinstance(choices, list) or len(choices) != 1:
+            return False
+
+        choice = choices[0]
+        if not isinstance(choice, dict) or set(choice.keys()) != {
+            "delta",
+            "index",
+            "finish_reason",
+        }:
+            return False
```
Comment (adubovik) [inline aidial_interceptors_sdk/examples/chat_completion/whitespace_accumulator.py]: Check that `choice["index"]` is equal to `0`.

File: aidial_interceptors_sdk/examples/chat_completion/whitespace_accumulator.py [hunk 30]
```diff
@@ -0,0 +1,114 @@
...
+            return None
+
+    def set_content(self, chunk: dict, content: str):
+        try:
+            chunk["choices"][0]["delta"]["content"] = content
+        except (KeyError, IndexError, TypeError):
+            return None
+
+    def set_chunk_if_needed(self, chunk: dict):
+        if not self.cur_chunk:
```
Comment (adubovik) [inline aidial_interceptors_sdk/examples/chat_completion/whitespace_accumulator.py]:
```py
self.cur_chunk = self.cur_chunk or chunk
```

File: aidial_interceptors_sdk/examples/chat_completion/whitespace_accumulator.py [hunk 31]
```diff
@@ -0,0 +1,114 @@
...
+        except (KeyError, IndexError, TypeError):
+            return None
+
+    def set_content(self, chunk: dict, content: str):
+        try:
+            chunk["choices"][0]["delta"]["content"] = content
+        except (KeyError, IndexError, TypeError):
+            return None
+
+    def set_chunk_if_needed(self, chunk: dict):
```
Comment (adubovik) [inline aidial_interceptors_sdk/examples/chat_completion/whitespace_accumulator.py]: Since this method is used only once, it could be inclined.

File: aidial_interceptors_sdk/examples/chat_completion/whitespace_accumulator.py [hunk 32]
```diff
@@ -0,0 +1,114 @@
...
+
+    def flush_if_needed(self) -> None:
+        if self.cur_chunk:
+            self.flush()
+
+    def need_flush(self, content: str | None) -> bool:
+        passed = time.perf_counter() - self.last_send
+        return content and not content.isspace() or (passed > self.timeout_sec)
+
+    def get_content(self, chunk: dict) -> str | None:
```
Comment (adubovik) [inline aidial_interceptors_sdk/examples/chat_completion/whitespace_accumulator.py]: `@staticmethod`

File: aidial_interceptors_sdk/examples/chat_completion/whitespace_accumulator.py (hunk 27 above)
Comment (adubovik) [inline aidial_interceptors_sdk/examples/chat_completion/whitespace_accumulator.py]: `@staticmethod`

File: aidial_interceptors_sdk/examples/chat_completion/whitespace_accumulator.py [hunk 33]
```diff
@@ -0,0 +1,114 @@
...
+    async def on_stream_chunk(self, chunk: dict) -> None:
+        if self.match_cache_structure(chunk):
+            self.set_chunk_if_needed(chunk)
+            content = self.get_content(chunk)
+            if content:
+                self.buffer.append(content)
+            if self.need_flush(content):
+                self.flush()
+            else:
+                self.cur_chunks += 1
```
Comment (adubovik) [inline aidial_interceptors_sdk/examples/chat_completion/whitespace_accumulator.py:93]: `content = None` leads to `self.cur_chunks += 1`

File: aidial_interceptors_sdk/examples/chat_completion/whitespace_accumulator.py [hunk 34]
```diff
@@ -0,0 +1,114 @@
...
+            "delta",
+            "index",
+            "finish_reason",
+        }:
+            return False
+
+        delta = choice["delta"]
+        if not isinstance(delta, dict) or not set(delta.keys()).issubset(
+            {
+                "content",
```
Comment (adubovik) [inline aidial_interceptors_sdk/examples/chat_completion/whitespace_accumulator.py]: We need to check that delta contains the only one key - "content" and that it's of type **string**.

File: tests/unit_tests/examples/test_whitespace.py [hunk 35]
```diff
@@ -0,0 +1,199 @@
...
+    if extra:
+        output["as_attachment"] = True
+    return output
+
+
+def extra(content: str | None = None, finish: bool = False) -> dict:
+    return wrap(content=content, finish=finish, extra=True)
+
+
+def has_content(chunk: dict) -> bool:
```
Comment (adubovik) [inline tests/unit_tests/examples/test_whitespace.py]: the method is unused

File: tests/unit_tests/examples/test_whitespace.py [hunk 36]
```diff
@@ -0,0 +1,199 @@
...
+            ),
+        )
+
+    else:
+        expected_chunks = check_stream(checker, interceptor_output)
+
+        for actual, expected in itertools.zip_longest(
+            response, expected_chunks
+        ):
+            print(f"\nexpected {expected}")
```
Comment (adubovik) [inline tests/unit_tests/examples/test_whitespace.py]: No need for debug prints in tests.

File: tests/unit_tests/examples/test_whitespace.py [hunk 37]
```diff
@@ -0,0 +1,199 @@
...
+from aidial_interceptors_sdk.examples.chat_completion import (
+    WhitespaceAccumulatorInterceptor,
+)
+from tests.utils.applications import WhitespaceEchoApplication
+from tests.utils.chunks import create_chunk_checker
+from tests.utils.dial_app import create_openai_client
+from tests.utils.json import has_type, match_objects, memorize
+
+
+def wrap(
```
Comment (adubovik) [inline tests/unit_tests/examples/test_whitespace.py:18]: Add a case when **both** `content` and `custom_content` are present.
Currently, it's either this or that.

Basically you need the method:

```py
def wrap(content: str | None = None, extra_content: str | None = None, finish: bool = False):
    ...
```

**Review by adubovik — CHANGES_REQUESTED** [review]

File: aidial_interceptors_sdk/examples/chat_completion/whitespace_accumulator.py [hunk 38]
```diff
@@ -0,0 +1,103 @@
...
+            chunk["choices"][0]["delta"]["content"] = content
+        except (KeyError, IndexError, TypeError):
+            return None
+
+    @staticmethod
+    def parse_content_only_chunk(chunk: dict) -> str | None:
+        if (choices := chunk.get("choices")) is None:
+            return None
+
+        choices = chunk.get("choices")
```
Comment (adubovik) [inline aidial_interceptors_sdk/examples/chat_completion/whitespace_accumulator.py]: redundant line

File: aidial_interceptors_sdk/examples/chat_completion/whitespace_accumulator.py [hunk 39]
```diff
@@ -0,0 +1,103 @@
...
+        _log.info(f"flushed {self.cur_chunks} chunks")
+        self.last_send = time.perf_counter()
+        content = "".join(self.buffer)
+        self.set_content(self.cur_chunk, content)
+        self.send_chunk(self.cur_chunk)
+        self.buffer.clear()
+        self.cur_chunk = None
+        self.cur_chunks = 0
+
+    def flush_if_needed(self) -> None:
```
Comment (adubovik) [inline aidial_interceptors_sdk/examples/chat_completion/whitespace_accumulator.py]: Let's remove this method and do the check inside `flush` instead:

```py
def flush(self) -> None:
        if self.cur_chunk is None:
            return
        ...
```

**Review by adubovik — CHANGES_REQUESTED** [review]

**Review by adubovik — APPROVED** [review]
